Extensions/MUC-Sub/XMPPMUCSub.m

```diff
@@ -0,0 +1,699 @@
+    // </iq>
+    
+    NSXMLElement *subscriptions = [NSXMLElement elementWithName:@"subscriptions"
+                                                          xmlns:XMPPMUCSubNamespace];
+    
+    NSString *iqId = [XMPPStream generateUUID];
+    
+    XMPPIQ *iq = [XMPPIQ iqWithType:@"get" elementID:iqId];
+    [iq addAttributeWithName:@"from" stringValue:xmppStream.myJID.bare];
+    [iq addAttributeWithName:@"to"   stringValue:domain];
+    
+    [iq addChild:subscriptions];
+    
+    [xmppIDTracker addElement:iq target:self selector:@selector(handleSubscriptionsAtQueryIQ:withInfo:) 
+                      timeout:60];
+    [xmppStream sendElement:iq];
+    
+    return iqId;
+}
+
+
+- (NSString *)subscribersIn:(XMPPJID *)room
+{
+    if (nil == room) {
+        return nil;
+    }
+    
+    // <iq from='hag66@shakespeare.example'
+    //       to='coven@muc.shakespeare.example'
+    //     type='get'
+    //       id='E6E10350-76CF-40C6-B91B-1EA08C332FC7'>
+    //   <subscriptions xmlns='urn:xmpp:mucsub:0' />
+    // </iq>
+    
+    NSXMLElement *subscriptions = [NSXMLElement elementWithName:@"subscriptions"
+                                                          xmlns:XMPPMUCSubNamespace];
+    
+    NSString *iqId = [XMPPStream generateUUID];
+    
+    XMPPIQ *iq = [XMPPIQ iqWithType:@"get" elementID:iqId];
+    [iq addAttributeWithName:@"from" stringValue:xmppStream.myJID.bare];
+    [iq addAttributeWithName:@"to"   stringValue:room.bare];
+    
+    [iq addChild:subscriptions];
+    
+    [xmppIDTracker addElement:iq target:self selector:@selector(handleSubscribersInQueryIQ:withInfo:) 
+                      timeout:60];
+    [xmppStream sendElement:iq];
+    
+    return iqId;
+}
+
+////////////////////////////////////////////////////////////////////////////////////////////////////
+#pragma mark -
+#pragma mark XMPPIDTracker
+////////////////////////////////////////////////////////////////////////////////////////////////////
+
+- (void)handleSubscribeQueryIQ:(XMPPIQ *)iq withInfo:(XMPPBasicTrackingInfo *)basicTrackingInfo
+{
+    dispatch_block_t block = ^{ @autoreleasepool {
+        NSXMLElement *query = [iq elementForName:@"subscribe" xmlns:XMPPMUCSubNamespace];
+        if (nil == query) {
+            // Can this actually happen? Still, safeguard for my conscience.
+            return;
+        }
+        
+        // "to" and "from" are the other way around in the response. Therefore the function
+        // call reads a bit funny. "to" is the receiving user and "from" the sending room.
+        if (iq.isResultIQ) {
+            [multicastDelegate xmppMUCSub:self didSubscribeUser:iq.to to:iq.from];
+        }
+        else {
+            [multicastDelegate xmppMUCSub:self didFailToSubscribe:iq.to 
+                                       to:iq.from 
+                                    error:[self errorFromIQ:iq]];
+        }
+    }};
+    
+    if (dispatch_get_specific(moduleQueueTag)) {
```

Comment:
What does this do? `true` if the current queue has label `moduleQueueTag`?

If so, not sure if executing `block` synchronously is the right thing to do. Older incoming "events" may be waiting on the queue?

Reply:
I copied this pattern from other modules. I noticed that XMPPIDTracker asserts that it is called on the correct queue. Can't really explain this construct.